README.md

[![Runboat](https://img.shields.io/badge/runboat-Try%20me-875A7B.png)](https://runboat.odoo-community.org/builds?repo=OCA/odoo-test-helper&target_branch=16.0)
[![Pre-commit Status](https://github.com/OCA/odoo-test-helper/actions/workflows/pre-commit.yml/badge.svg?branch=16.0)](https://github.com/OCA/odoo-test-helper/actions/workflows/pre-commit.yml?query=branch%3A16.0)
[![Build Status](https://github.com/OCA/odoo-test-helper/actions/workflows/test.yml/badge.svg?branch=16.0)](https://github.com/OCA/odoo-test-helper/actions/workflows/test.yml?query=branch%3A16.0)
[![codecov](https://codecov.io/gh/OCA/odoo-test-helper/branch/16.0/graph/badge.svg)](https://codecov.io/gh/OCA/odoo-test-helper)
[![Translation Status](https://translation.odoo-community.org/widgets/odoo-test-helper-16-0/-/svg-badge.svg)](https://translation.odoo-community.org/engage/odoo-test-helper-16-0/?utm_source=widget)

<!-- /!\ do not modify above this line -->

# odoo-test-helper

None

<!-- /!\ do not modify below this line -->

<!-- prettier-ignore-start -->

[//]: # (addons)

This part will be replaced when running the oca-gen-addons-table script from OCA/maintainer-tools.

[//]: # (end addons)

<!-- prettier-ignore-end -->

## Licenses

This repository is licensed under [AGPL-3.0](LICENSE).

However, each module can have a totally different license, as long as they adhere to Odoo Community Association (OCA)
policy. Consult each module's `__manifest__.py` file, which contains a `license` key
that explains its license.

----
OCA, or the [Odoo Community Association](http://odoo-community.org/), is a nonprofit
organization whose mission is to support the collaborative development of Odoo features
and promote its widespread use.


CHANGES.rst
Changes
~~~~~~~

.. Future (?)
.. ----------
.. - ...

2.1.1
-----

[FIX] FakeModelLoader: avoid loading module again on restore_registry() for Odoo 15.0+
[FIX] Ensures registy is cleaned up in addons without models


2.1.0
-----

- [IMP] Allow to ignore Odoo core modules to avoid warning

2.0.5
-----

- .gitignore added

2.0.4
-----

- [FIX] AttributeError in Odoo 15+, regression introduced in 2.0.3

2.0.3
-----

- [FIX] restore_registry in Odoo 16

    See odoo/odoo@cd12293

    This new attribute is the source of truth for the base classes
    and in setup_models (called further down in the modified code in this PR),
    the model's base classes are reset from it:

    https://github.com/odoo/odoo/blob/e1f06479a526c703ccabc441b1e194646206b966/odoo/models.py#L2728-L2730.

    The test failure fixed by this PR can be inspected in
    https://app.travis-ci.com/github/OCA/odoo-test-helper/builds/258453331


2.0.2
-----

- Fix ``mock`` import for v15

2.0.1
-----

- Fix support for Odoo 15.0


2.0.0
-----

- Move to OCA
- Re-license to LGPL


1.1.0
-----

- Refactoring (misc imp/fix)


1.0.0
-----

- Initial release


tests/scripts/install_odoo.py
#!/usr/bin/env python
import subprocess
import sys

odoo_branch = sys.argv[1]


def odoo_installed():
    try:
        import odoo  # noqa

        return True
    except ImportError:
        # odoo < 10
        try:
            import openerp  # noqa

            return True
        except ImportError:
            # odoo not installed
            return False


def install_odoo():
    if odoo_branch in ["10.0", "11.0", "12.0", "13.0"]:
        # setuptools 58 dropped support for 2to3, which is required
        # for dependencies of older Odoo versions
        subprocess.check_call(
            [
                "pip",
                "install",
                "setuptools<58",
            ]
        )
    url = f"https://github.com/odoo/odoo/archive/refs/heads/{odoo_branch}.tar.gz"
    req_url = (
        f"https://raw.githubusercontent.com/odoo/odoo/{odoo_branch}/requirements.txt"
    )
    subprocess.check_call(
        [
            "pip",
            "install",
            url,
        ]
    )
    subprocess.check_call(
        [
            "pip",
            "install",
            "-r",
            req_url,
        ]
    )


def main():
    if not odoo_installed():
        install_odoo()
    return sys.exit(0)


if __name__ == "__main__":
    main()


odoo_test_helper/fake_model_loader.py
# -*- coding: utf-8 -*-
# Copyright 2018 ACSONE (http://www.acsone.eu).
# @author: Laurent Mignon <laurent.mignon@acsone.eu>
# Copyright 2018 Camptocamp SA (http://www.camptocamp.com).
# @author: Simone Orsi <simone.orsi@camptocamp.com>
# Copyright 2020 Akretion (http://www.akretion.com).
# @author: Sébastien BEAU <sebastien.beau@akretion.com>
# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).

import logging

from odoo import models
from odoo.release import version_info
from odoo.tools import OrderedSet

try:
    from unittest import mock
except ImportError:
    import mock

from .whitelist import WHITELIST

module_to_models = models.MetaModel.module_to_models
_logger = logging.getLogger(__name__)


class FakePackage(object):  # noqa
    def __init__(self, name):
        self.name = name


class FakeModelLoader(object):
    """Loader for Odoo tests fake models.

    Usage on SavepointTestCase:

        from odoo_test_helper import FakeModelLoader

        @classmethod
        def setUpClass(cls):
            super().setUpClass()
            cls.loader = FakeModelLoader(cls.env, cls.__module__)
            cls.loader.backup_registry()
            from .fake_models import MyFakeModelClass1, MyFakeModelClass2
            cls.loader.update_registry(
                (MyFakeModelClass1, MyFakeModelClass2)
            )

        @classmethod
        def tearDownClass(cls):
            cls.loader.restore_registry()
            super().tearDownClass()

    Usage on TransactionCase / HttpCase:

        from odoo_test_helper import FakeModelLoader

        def setUp(self):
            super().setUp()
            self.loader = FakeModelLoader(self.env, self.__module__)
            self.loader.backup_registry()
            from .fake_models import MyFakeModelClass1, MyFakeModelClass2
            self.loader.update_registry(
                (MyFakeModelClass1, MyFakeModelClass2)
            )

        def tearDown(self):
            self.loader.restore_registry()
            super().tearDown()
    """

    _original_registry = None
    _original_module2modules = None
    _module_name = None

    def __init__(self, env, __module__):
        self.env = env
        if hasattr(self.env.registry["base"], "_get_addon_name"):
            # Only before V14
            self._module_name = self.env.registry["base"]._get_addon_name(__module__)
        else:
            self._module_name = __module__.split(".")[2]

    def _check_wrong_import(self):
        version = str(version_info[0])
        modules_to_ignore = WHITELIST.get(version, [])
        for module, odoo_models in module_to_models.items():
            for model in odoo_models:
                module_path = model.__module__
                path = module_path.split(".")
                if module_path not in modules_to_ignore and path[3] == "tests":
                    _logger.warning(
                        "Wrong Import in module {}, the class {} have been already "
                        "imported.\nPlease take a look to the README on how to "
                        "import test class".format(module, model)
                    )

    def backup_registry(self):
        self._check_wrong_import()
        self._original_registry = {}
        self._original_module_to_models = {}
        for model_name, model in self.env.registry.models.items():
            self._original_registry[model_name] = {
                "base": model.__bases__,
                "_fields": model._fields.copy(),
                "_inherit_children": OrderedSet(model._inherit_children._map.keys()),
                "_inherits_children": set(model._inherits_children),
            }
        for key in module_to_models:
            self._original_module_to_models[key] = list(module_to_models[key])

    def _clean_module_to_model(self):
        for key in self._original_module_to_models:
            module_to_models[key] = list(self._original_module_to_models[key])
        for key in set(module_to_models.keys()).difference(
            self._original_module_to_models.keys()
        ):
            # remove module that have been added during the test
            del module_to_models[key]

    def update_registry(self, odoo_models):
        # Since V13 field are computed at the end
        # In the setup of your test if you create or modify a record that
        # need to be recomputed we need to recompute them before reloading
        # the registry
        if hasattr(self.env, "flush_all"):  # Odoo 16 and later
            self.env.flush_all()
        elif hasattr(self.env.all, "tocompute"):
            to_recompute_models = set()
            for field, _vals in self.env.all.tocompute.items():
                to_recompute_models.add(field.model_name)
            for model in to_recompute_models:
                self.env[model].recompute()

        # In case that you are re-using fake model from an other module
        # Odoo have already modify the module_to_models variable
        # It updated everytime we import a python file
        # So we need first to clean this variable
        # then we can inject the fake class in the module_to_models
        # so we can reload the registry correctly with only the given class
        self._clean_module_to_model()
        for model in odoo_models:
            if model not in module_to_models[self._module_name]:
                module_to_models[self._module_name].append(model)

        with mock.patch.object(self.env.cr, "commit"):
            model_names = self.env.registry.load(
                self.env.cr, FakePackage(self._module_name)
            )
            self.env.registry.setup_models(self.env.cr)
            self.env.registry.init_models(
                self.env.cr, model_names, {"module": self._module_name}
            )

    def restore_registry(self):
        for key in self._original_registry:
            ori = self._original_registry[key]
            model = self.env.registry[key]
            if hasattr(model, "_BaseModel__base_classes"):  # As of 15.0
                model._BaseModel__base_classes = ori["base"]
            model.__bases__ = ori["base"]
            model._inherit_children = ori["_inherit_children"]
            model._inherits_children = ori["_inherits_children"]
            for field in model._fields:
                if field not in ori["_fields"]:
                    if hasattr(model, field):
                        delattr(model, field)
            model._fields = ori["_fields"]

        # delete 1st models w/out children
        sorted_models = sorted(
            self.env.registry.models.items(), key=lambda x: x[1]._inherit_children
        )
        for name, __ in sorted_models:
            if name not in self._original_registry:
                del self.env.registry.models[name]

        self._clean_module_to_model()

        # reload is needed to reset correctly the field on the record
        # but only up to Odoo 14
        if version_info[0] < 15:

            with mock.patch.object(self.env.cr, "commit"):
                self.env.registry.model_cache.clear()
                model_names = self.env.registry.load(
                    self.env.cr, FakePackage(self._module_name)
                )
                self.env.registry.setup_models(self.env.cr)
                self.env.registry.init_models(
                    self.env.cr, model_names, {"module": self._module_name}
                )
